server/lib/realtime/replication.ex

```diff
@@ -209,83 +208,140 @@ defmodule Realtime.Replication do
+          commit_timestamp: commit_timestamp
+        }
 
-    updated_record = %UpdatedRecord{
-      type: "UPDATE",
-      schema: relation.namespace,
-      table: relation.name,
-      columns: relation.columns,
-      old_record: old_data,
-      record: data
-    }
+        %State{
+          state
+          | transaction: {lsn, %{txn | changes: [updated_record | changes]}}
+        }
 
-    {lsn, txn} = state.transaction
-    %{state | transaction: {lsn, %{txn | changes: [updated_record | txn.changes]}}}
+      _ ->
+        state
+    end
   end
 
-  defp process_message(%Delete{} = msg, state) do
-    relation = Map.get(state.relations, msg.relation_id)
-
-    data =
-      data_tuple_to_map(
-        relation.columns,
-        msg.old_tuple_data || msg.changed_key_tuple_data
-      )
+  defp process_message(
+         %Delete{
+           relation_id: relation_id,
+           old_tuple_data: old_tuple_data,
+           changed_key_tuple_data: changed_key_tuple_data
+         },
+         %State{
+           relations: relations,
+           transaction: {lsn, %{commit_timestamp: commit_timestamp, changes: changes} = txn}
+         } = state
+       )
+       when is_map(relations) do
+    case Map.fetch(relations, relation_id) do
+      {:ok, %{columns: columns, namespace: namespace, name: name}} when is_list(columns) ->
+        data = data_tuple_to_map(columns, old_tuple_data || changed_key_tuple_data)
+
+        deleted_record = %DeletedRecord{
+          type: "DELETE",
+          schema: namespace,
+          table: name,
+          columns: columns,
+          old_record: data,
+          commit_timestamp: commit_timestamp
+        }
 
-    deleted_record = %DeletedRecord{
-      type: "DELETE",
-      schema: relation.namespace,
-      table: relation.name,
-      columns: relation.columns,
-      old_record: data
-    }
+        %State{state | transaction: {lsn, %{txn | changes: [deleted_record | changes]}}}
 
-    {lsn, txn} = state.transaction
-    %{state | transaction: {lsn, %{txn | changes: [deleted_record | txn.changes]}}}
+      _ ->
+        state
+    end
   end
 
-  defp process_message(%Truncate{} = msg, state) do
-    truncated_relations =
-      for truncated_relation <- msg.truncated_relations do
-        relation = Map.get(state.relations, truncated_relation)
-
-        %TruncatedRelation{
-          type: "TRUNCATE",
-          schema: relation.namespace,
-          table: relation.name
-        }
-      end
-
-    {lsn, txn} = state.transaction
+  defp process_message(
+         %Truncate{truncated_relations: truncated_relations},
+         %State{
+           relations: relations,
+           transaction: {lsn, %{commit_timestamp: commit_timestamp, changes: changes} = txn}
+         } = state
+       )
+       when is_list(truncated_relations) and is_list(changes) and is_map(relations) do
+    new_changes =
+      Enum.reduce(truncated_relations, changes, fn truncated_relation, acc ->
```

Comment:
@soedirgo I know we talked about this and you decided to go with the [imperative code](https://github.com/supabase/realtime/pull/101#discussion_r556978624) when I suggested the [further optimization](https://github.com/supabase/realtime/pull/101#discussion_r556927646). As I was refactoring, I found that using `reduce` encapsulated the more complex optimization logic from my suggestion while making it clearer.

I don't mean to undermine what you decided to go with so more than happy to revert the changes here. Let me know!

Reply:
👍 this looks good to me actually, I was mostly worried about introducing extra helper functions.